Prevent unnecessary verification of parity blocks while reading

fwessels opened the pull request:
While reading an erasure coded file, in case any of the data blocks is missing and we need to reconstruct data, we are unnecessarily verifying the integrity of the parity blocks.

As bit-rot protection shields us from any changes in either the data or parity blocks, it is pointless to verify the integrity of any existing (not reconstructed) parity blocks as these cannot be incorrect (or we would have a bug in the galois field arithmetic or matrix manipulation...). 

Moreover we are not using the parity blocks while reading an erasure coded file, as the data blocks represent the contents of the file.
 
Note that this is not the case for healing where we would also want to reconstruct any missing parity blocks.

Furthermore added a TODO to add a data block reconstruction only mode to Reed Solomon as a further optimization.

## Motivation and Context
Performance optimization by preventing unnecessary verification.

## Types of changes
- [x] New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code follows the code style of this project.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

fwessels committed:
Prevent unnecessary verification of parity blocks while reading erasure coded file.

donatello commented on cmd/erasure_test.go:
```diff
@@ -124,26 +124,54 @@ func TestErasureDecode(t *testing.T) {
+					err := errUnexpected
+					t.Errorf("Test %d: Expected to pass by failed instead %s", i+1, err)
+				}
 			}
```
Nice style of using blocks here to be able to cleanly redeclare and reuse variable names. I will use this in tests in the future! :+1:

donatello reviewed (approved):
LGTM - please take decision to merge after considering @aead's comment.

fwessels committed:
Merge branch 'master' of https://github.com/minio/minio into feature/prevent-parity-verify

fwessels committed:
Update klauspost/reedsolomon and just only reconstruct data blocks while reading (prevent unnecessary parity block reconstruction)

harshavardhana commented on cmd/erasure-readfile.go:
```diff
@@ -328,14 +340,23 @@ func decodeData(enBlocks [][]byte, dataBlocks, parityBlocks int) error {
+		//
+		// Note that in the presence of bit-rot protection (as is present in Minio)
+		// you really should never see this message.
+		err = errors.New("Unable to verify parity blocks, either existing data or parity blocks corrupted")
```
We can perhaps add a test case for this.. just for the failure case of Verify() ? @fwessels

fwessels replied:
Discuss this and decided to remove the Verification altogether since in our case we are protected by bit rot protection.

fwessels committed:
Remove Verification of (all) reconstructed Data and Parity blocks since in our case we are protected by bit rot protection.
And even if the Verification would fail (essentially impossible) there is no way to definitively say whether the data is still correct or not, so this call make no sense for our use case.

deekoder committed:
Merge branch 'master' into feature/prevent-parity-verify
